### PR TITLE
Fermion force

### PR DESCRIPTION
**Pull request by sy3394**

### CODE REVIEW

File: test/test_gauge_field.py [hunk 1]
```diff
@@ -227,6 +227,38 @@ def test_force(lib, lattice, device, epsilon):
...
+
+    R = spinor(lattice, dtype=dtype)
+    R.gaussian()
+
+    # U'- U ~ eps*mom where U' = exp(eps*mom)*U
+    for parity in [None, "EVEN"]:
+        params = {"kappa":0.01, "csw":1}
+        phi = gf.Dirac(**params).Mdag(R)
+        action = gf.Dirac(**params).action(phi, parity=parity) 
+        action2 = gf2.Dirac(**params).action(phi, parity=parity)
```
Comment (sbacchio) [inline test/test_gauge_field.py]: Define
```
D = gf.Dirac(**params)
D2 = gf2.Dirac(**params)
```
and use those

File: lyncs_quda/gauge_field.py [hunk 2]
```diff
@@ -16,13 +16,16 @@
...
+from array import array
 import numpy
 from lyncs_cppyy import make_shared, lib as tmp, to_pointer, array_to_pointers
 from lyncs_utils import prod, isiterable
 from .lib import lib, cupy
 from .lattice_field import LatticeField, backend
+from .spinor_field import spinor
 from .time_profile import default_profiler
 
+import lyncs_cppyy
```
Comment (sbacchio) [inline lyncs_quda/gauge_field.py]: The added imports above, I guess are not needed anymor

File: lyncs_quda/dirac.py [hunk 3]
```diff
@@ -33,31 +32,45 @@ class Dirac:
     csw: float = 0
     mu: float = 0
     epsilon: float = 0
+    matPCtype: str = "INVALID"
```
Comment (sbacchio) [inline lyncs_quda/dirac.py]: does this parameter need to be exposed here?
Maybe better to have `parity` as a parameter here and then set `matPCtype` accordingly?
I guess then you don't need to change the attributes if parity is fixed for the operator

Comment (sy3394) [reply lyncs_quda/dirac.py]: there are 4 possibilities (EVEN and ODD with SYMM or ASYMM)

File: test/test_lib.py [hunk 4]
```diff
@@ -12,3 +12,7 @@ def test_init(lib):
     assert not lib.initialized
     lib.init_quda()
     assert lib.initialized
+
+
+def test_excep(lib):
+    lib.test_exception()
```
Comment (sbacchio) [inline test/test_lib.py]: I think this shouldn't have been commiter, or?

File: lyncs_quda/dirac.py [hunk 5]
```diff
@@ -33,31 +32,48 @@ class Dirac:
     csw: float = 0
     mu: float = 0
     epsilon: float = 0
+    full: int = field(default=True)
```
Comment (sbacchio) [inline lyncs_quda/dirac.py]: Do you need field? `= True` should be enough
And use `bool` instead of `int` as type

**Review by sbacchio — APPROVED** [review]
Very good, I like it a lot!
So just two trivial comments to resolve and can get in

**Pull request merged**